File: CountSubarraysXOR/count_subarr_k_xor.cpp
#include<iostream>
#include<unordered_map>
#include<vector>
using namespace std;
// Function to count subarrays with given XOR
int countSubarraysWithXOR(const vector<int>& arr, int k) {
    unordered_map<int, int> prefixXORCount;
    int count = 0;
    int currentXOR = 0;

    for (int num : arr) {
        currentXOR ^= num;

        // If currentXOR is equal to k, increment count
        if (currentXOR == k) {
            count++;
        }

        // If (currentXOR ^ k) exists in the map, add its count to result
        if (prefixXORCount.find(currentXOR ^ k) != prefixXORCount.end()) {
            count += prefixXORCount[currentXOR ^ k];
        }

        // Increment the count of currentXOR in the map
        prefixXORCount[currentXOR]++;
    }

    return count;
}
int main() {
    vector<int> arr = {4, 2, 2, 6, 4};
    int k = 6;
    cout << "Count of subarrays with XOR " << k << " is: " << countSubarraysWithXOR(arr, k) << endl;
    return 0;
}
// Output: Count of subarrays with XOR 6 is: 4

File: CountSubarraysXOR/count_subarr_k_xor.rs
//use std::collections::HashMap;
// Function to count the number of subarrays with a given XOR
fn count_sub_arrays_xor(arr: Vec<i32>, k: i32) -> i32 {
    let mut count = 0;
    let mut xor_map = std::collections::HashMap::new();
    let mut current_xor = 0;

    for &num in arr.iter() {
        current_xor ^= num;

        if current_xor == k {
            count += 1;
        }

        if let Some(&prev_count) = xor_map.get(&(current_xor ^ k)) {
            count += prev_count;
        }

        *xor_map.entry(current_xor).or_insert(0) += 1;
    }

    count
}
fn main() {
    let arr = vec![4, 2, 2, 6, 4];
    let k = 6;
    let result = count_sub_arrays_xor(arr, k);
    println!("Count of subarrays with XOR {}: {}", k, result);
}